File: 07_HashTables/Open Addressing/Open Add CPP/OpenAdd.cpp
#include "OpenAdd.hpp"

openAddHash::openAddHash(int v=0)
{
        cap = 20;
        N = v;  
        table.resize(cap);
}

void openAddHash::printHash()
{

        //printing vector of vectors
        for(int i = 0; i < table.size(); ++i)
        {
                for(int j = 0; j < table.at(i).size(); ++j)
                {
                        cout << i << "---->";
                        cout << table[i][j];
                }
                cout<<endl;
        }

}

int openAddHash::hashFunc(int x)
{
        return x % cap;
}

void openAddHash::insertVal(int value)
{
        int key = hashFunc(value);

        if(table[key].empty())
        {
                table[key].push_back(value);
                return;
        }

        int newIdx = (key + 1) % cap;
        while(newIdx != key && !table[newIdx].empty())
        {
                newIdx = (newIdx + 1) % cap;
        }

        if(newIdx == key)
        {
                cout<<"Hash Full\n";
                return;
        }

        table[newIdx].push_back(value);
}

void openAddHash::deleteVal(int target)
{
    int key = hashFunc(target);

    for (int i = 0; i < table[key].size(); ++i)
    {
        if (table[key][i] == target)
        {
            table[key].erase(table[key].begin() + i);
            cout << "Deleted " << target << endl;
            return;
        }
    }

    cout << target << " not found\n";
}

bool openAddHash::ifPresent(int target)
{
    int key = hashFunc(target);

    for (int i = 0; i < table[key].size(); ++i)
    {
        if (table[key][i] == target)
            return true;
    }

    return false;
}


int main()
{
        vector<int>nums = {12,33,45,67,77,57,34,32,63,91,40};

        openAddHash obj(10);

        for( int i = 0; i < nums.size(); ++i)
        {
                obj.insertVal(nums.at(i));
        }

        cout << "Hash Table"<<endl;
        obj.printHash();

        cout << "Delete 77 from table\n";
        obj.deleteVal(77);
        obj.printHash();

        cout << "check if 33 is present\n";
        if(obj.ifPresent(33))
        {
                cout<<"Found\n";
        }


        return 0;
}

File: 07_HashTables/Open Addressing/Open Add Java/OpenAdd.java
import java.io.IOException;
import java.util.*;

class OpenAdd
{
        int N;
        int cap;
        ArrayList<Integer>[] table;

        public OpenAdd()
        {
                cap = 20;
                N= 0;
        }

        public OpenAdd(int x)
        {
                N= x;
                cap = 20;
                table = new ArrayList[cap];
                for(int i = 0; i<cap; ++i)
                {
                        table[i] = new ArrayList<>();
                }
        }

        public int hashFunc(int x)
        {
                return x % cap;
        }

        public void insertVal(int val)
        {
                int idx = hashFunc(val);
                if(table[idx].isEmpty())
                {
                        table[idx].add(val);
                        return;
                }

                int newIdx = (idx + 1) % cap;
                while(newIdx != idx && !table[newIdx].isEmpty())
                {
                        newIdx = (newIdx + 1) % cap;
                }

                while(newIdx == idx)
                {
                        System.out.println("Hash Full");
                        return;
                }

                table[newIdx].add(val);
                
        }

        public void deleteVal(int target)
        {
                int idx = hashFunc(target);
                for(int i = 0; i < table[idx].size(); ++i)
                {
                        if(table[idx].get(i) == target)
                        {
                                table[idx].remove(i);
                                System.out.println("Deleted" + target);
                                return;
                        }
                }
                System.out.println("Not Found");
        }

        public boolean ifPresent(int target)
        {
                int idx = hashFunc(target);
                for(int i = 0; i < table[idx].size(); ++i)
                {
                        if(table[idx].get(i) == target)
                        {
                                return true;
                        }
                }
                System.out.println("Not Found");
                return false;
        }

        public void printHash()
        {
                System.out.println("Hash Table");
                for(int i = 0; i != cap; ++i)
                {
                        for(int j = 0; j < table[i].size(); ++j)
                        {
                                System.out.print(i + "---->");

                                System.out.print(table[i].get(j));
                        }
                        System.out.println();
                }
        }

        public static void main(String[] args) throws IOException 
        {
            int[] nums = {12,33,45,67,77,57,34,32,63,91,40};

            OpenAdd obj = new OpenAdd(10);

            for( int i = 0; i < nums.length; ++i)
            {
                obj.insertVal(nums[i]);
            }

            System.out.println("Hash Table");
            obj.printHash();

            System.out.println("Delete 77 from table");
            obj.deleteVal(77);
            obj.printHash();

            System.out.println("CHeck if 33 is present");
            if(obj.ifPresent(33))
            {
                System.out.println("Found");
            }
        }


}